### PR TITLE
Add `output` option to `build` and `dist-dir` option to `publish` command

### PR DESCRIPTION
**Pull request by benjamin-callonnec**
# Pull Request Check List

- Add `output` option to `build` command.
- Add `dist-dir` option to `publish` command.

Resolves: #3586 

Similar PR: #7208 

- [x] Added **tests** for changed code.
- [x] Updated **documentation** for changed code.

### CODE REVIEW

**Comment by Secrus** [comment]
This also requires changes to the publishing setup, since using custom build output would make publishing the built artifacts with Poetry impossible.

**Comment by benjamin-callonnec** [comment]
Thank for the comment @Secrus, I tried to take it into consideration!

File: src/poetry/publishing/uploader.py [hunk 1]
```diff
@@ -61,9 +61,21 @@ def user_agent(self) -> str:
         agent: str = user_agent("poetry", __version__)
         return agent
 
+    @property
+    def dist_dir(self) -> Path:
+        io_dist_dir = "dist"
+        if "dist-dir" in self._io.input.options:  # Option comes with publish command.
```
Comment (Secrus) [inline src/poetry/publishing/uploader.py:67]: I think it would be better to put dist dir as an optional init argument for Uploader.

Comment (radoering) [reply src/poetry/publishing/uploader.py:67]: Yeah, I suppose we want to try to access command line options only in commands (via `self.option()`) and pass them explicity to classes that need them.

Comment (benjamin-callonnec) [reply src/poetry/publishing/uploader.py]: What about now?

File: tests/console/commands/test_build.py [hunk 2]
```diff
@@ -16,11 +18,15 @@
     from tests.types import FixtureDirGetter
 
 
-def test_build_with_multiple_readme_files(
+@pytest.mark.parametrize(
+    "output_dir", [None, "dist", "test/dir", "../dist", "absolute"]
+)
+def test_build_with_multiple_readme_files_and_output_option(
```
Comment (radoering) [inline tests/console/commands/test_build.py:24]: I'd prefer to leave this test alone and add a new test. You could use `test_build_creates_packages_in_dist_directory_if_no_output_is_specified` from #8877 as starting point (and copy the relevant stuff from there).

Comment (benjamin-callonnec) [reply tests/console/commands/test_build.py]: @radoering done!

**Comment by Secrus** [comment]
Could you rebase and resolve conflicts?

**Comment by benjamin-callonnec** [comment]
> Could you rebase and resolve conflicts?

Done!

File: src/poetry/console/commands/build.py [hunk 3]
```diff
@@ -2,18 +2,12 @@
...
-from typing import TYPE_CHECKING
-
 from cleo.helpers import option
 
 from poetry.console.commands.env_command import EnvCommand
 from poetry.utils.env import build_environment
 
 
-if TYPE_CHECKING:
-    from pathlib import Path
```
Comment (benjamin-callonnec) [inline src/poetry/console/commands/build.py:12]: Type checking is not allowed because is used not for type checking line 60

**Review by Secrus — APPROVED** [review]
LGTM. @radoering could you also take a look?

File: src/poetry/console/commands/build.py [hunk 4]
```diff
@@ -1,23 +1,26 @@
...
     name = "build"
     description = "Builds a package, as a tarball and a wheel by default."
 
     options = [
-        option("format", "f", "Limit the format to either sdist or wheel.", flag=False)
+        option("format", "f", "Limit the format to either sdist or wheel.", flag=False),
+        option(
+            "output",
+            "o",
+            "Set output directory for build artifacts. Default is `dist`.",
```
Comment (radoering) [inline src/poetry/console/commands/build.py:20]: Normally, we don't use backticks here (only in the docs), but I do not have strong feelings and did not found a reference for something else so 🤷

File: tests/console/commands/test_publish.py [hunk 5]
```diff
@@ -130,3 +136,77 @@ def test_skip_existing_output(
...
+    http: type[httpretty.httpretty],
+    fixture_dir: FixtureDirGetter,
+    tmp_path: Path,
+    tmp_venv: VirtualEnv,
+    command_tester_factory: CommandTesterFactory,
+    dist_dir: str | None,
+) -> None:
+    source_dir = fixture_dir("with_multiple_dist_dir")
+    target_dir = tmp_path / "project"
+    shutil.copytree(str(source_dir), str(target_dir))
```
[the remark this answers is not in the document]
Comment (benjamin-callonnec) [reply tests/console/commands/test_publish.py:189]: Done, you can resolve the conversation

**Review by radoering — APPROVED** [review]

**Pull request merged**